### ISSUE TITLE
php-mlx architecture

### ISSUE THREAD

**Issue by akondas** [php-ai/php-mlx#1]
Because it is a new library, we now have a moment to discuss architecture. php-ml in its implementation as a dataset used ordinary native arrays. This was associated with a large amount of memory loss. I would like to focus mainly on the architecture of datasets and its circulation among estimators, transformers, etc.

**Comment by SerafimArts** [comment]
@akondas hi! What do you think about the use of computational shaders to speed up work?

I can provide OpenCL or Vulkan API (like this: https://github.com/BicEngine/Vulkan) for interacting with GPU

**Comment by akondas** [comment]
Absolutely. I have already experimented with cuda bindings, but I can't find this library right now. I suggest you do a separate issue :wink:

**Comment by ideaguy3d** [comment]
Would it be possible / efficient to use "php-ds" or the spl data structures as an alternative? 

I refactored an app I built from arrays to php-ds vectors and decreased memory by ~50%, However, going from 2GB of mem to 1GB was still too much so I resorted to streams (which maintained ~19MB of mem usage during runtime).

**Comment by sencere** [comment]
I would like to see a random forest implementation :heart: I'm also open for contributing to this project.

**Comment by hamidrezabstn** [comment]
Guys ! with FFI in our door steps , why dont we jump for highest ? 
i think php can be useful as python if we take python approach .
using php as an interface for huge computation

**Comment by KEINOS** [comment]
@hamidrezabstn 

> i think php can be useful as python

Hell yeah! 

And not only for the interface, I believe.

The [upcoming PHP8](https://github.com/php/php-src/releases) major release, with the JIT enabled, is much much faster than I expected.

Some scripts are faster than Golang. And believe or not [PHP8 with JIT enabled in Docker](https://github.com/KEINOS/Dockerfile_of_PHP8-JIT) on Raspberry Pi3 is **faster** than running local on macOS' default PHP 7.3.11! (Thanks for [@akondas' blog article](https://arkadiuszkondas.com/compiling-php-8-from-source-with-jit-support/))

I believe from PHP 8.0 things will change and bring a recursive acronym escapade of Programmers Hate PHP.

The only problem is that I am no good at math at all ...  Σ(￣□￣；）What?!

**Comment by KEINOS** [comment]
> architecture of datasets

I prefer SQLite3 as a storage of the datasets since it's simple, lightweight, stable, and fast. But what about `.h5` format?

I think many people use [HDF5](https://en.wikipedia.org/wiki/Hierarchical_Data_Format#HDF5) format datasets. So, it would be handy if we can use them by default, wouldn't it?

Something like [`h5py`](https://docs.h5py.org/en/stable/high/dataset.html) might help. But so far it seems there's no [package for HDF5](https://packagist.org/?query=HDF5) nor something like [`h5php`](https://www.google.com/search?q=HDF5+h5php&oq=HDF5+h5php).

The below was the only discussion I found about HDF5 in PHP. 

- [How to read HDF5 files in a PHP script]() @ Quora

However, it didn't reach to the answer rather than switching to Python. The main reason seems to be by not having an extension for it.

Since there's a Clang source, as a theory, an extension might be created. Yea, as a theory...

- [Introduction to HDF5](https://support.hdfgroup.org/HDF5/doc/H5.intro.html) @ hdfgroup.org
- [Clang](https://support.hdfgroup.org/HDF5/examples/intro.html#c) | HDF5 INTRODUCTORY EXAMPLES @ hdfgroup.org

**Comment by akondas** [comment]
Storage is only a implementation details. And as we know, the devil is in the details :wink: 

There are two more topics left for me to close and I will be able to go back to the php-mlx implementation. I am assuming that particular algorithms (classifiers or regressors) should be interfaces and have different implementations. This will allow the use of an FFI, for example. I am eager to see a PoC solution like this, to be able to state whether it makes sense in general.

In addition, I would also like to focus more on the dataset itself. Enable operations such as the Pandas library and the visualization method like Jupyter Notebooks. 

I have also described my main concerns in this article: https://arkadiuszkondas.com/3-reasons-why-php-is-not-yet-perfect-for-machine-learning/

If someone wants to start now, I am happy to help.

**Comment by KEINOS** [comment]
> particular algorithms (classifiers or regressors) should be interfaces 

So, you mean to define something like an API first so to be able to switch the backend? That makes sense.

>  I would also like to focus more on the dataset itself. Enable operations such as the Pandas library and the visualization method like Jupyter Notebooks.

Since I wasn't familiar with Jupyter Notebook, I implemented [PHP8-RC2 Kernel to work on Jupyter](https://github.com/KEINOS/Dockerfile_of_PHP8-JIT/tree/php8-jit/samples/jupyter-php) to see how they work. And yea, I feel there is a lack of visualization method in PHP. Something like `matplotlib` we need as you mentioned the [article](https://arkadiuszkondas.com/3-reasons-why-php-is-not-yet-perfect-for-machine-learning/).

Is it reasonable to create something like `matplotlib` in pure PHP rather than make a parasitical wrapper of Python's `matplotlib` in PHP?

**Comment by yuichiis** [comment]
>> https://arkadiuszkondas.com/3-reasons-why-php-is-not-yet-perfect-for-machine-learning/
I read the above article. I strongly agree.

I have already made it, except for the Jupyter.

- https://rindow.github.io/mathematics/matrix/matrix.html
- https://rindow.github.io/mathematics/plot/overviewplot.html
- https://rindow.github.io/mathematics/openblas/overviewopenblas.html
- https://rindow.github.io/mathematics/acceleration/opencl.html

**Comment by akondas** [comment]
Looks very promising. Thanks to this job there is some slight chance of entering the mainstream. But you will have to put in a lot of work that no one pays for.

Unfortunately, comments like "abandoned and disputed php-ml" (https://github.com/nextcloud/suspicious_login/pull/375) do not encourage me to further work on this library.

I am happy to accept / supervise the work (PR's), but in the near future I am not going to write the code myself, so far I have other interesting things to do.

**Comment by KEINOS** [comment]
@akondas 

I'm sorry to hear that. I understand that anyone can lose one's drive, after working so hard all alone, and taken all down in no time by the subsequent projects' DMCA attack. I felt unfair with the rashness of them and DMCA.

No matter what they say, it's as concrete as anything else that PHP-ML was the pioneer of machine learning in PHP!

Take your time, do what you want to do. I'm always on your side dude.

**Comment by yuichiis** [comment]
@akondas

I am sorry to hear your problem. The honor of the pioneer remains unchanged.
Do something fun for you. But I hope you will come back to write the program in the future.
I'm writing code because it's fun too.

My hope is the spread of interoperable interfaces.

I think that the data structure for high-speed numerical calculation that is aware of SIMD instructions and GPU is very important for PHP. I have the NDArray and LinearBuffer interfaces in the "interop-phpobjects" I'm using. That is the core of high-speed computing.
If a lot of compatible libraries and PHP extensions are created, we can expect a little game change for python. More people should write mathematical libraries in PHP.

The interface definition is independent of my library, so I aim to be able to replace and discard my library when other advanced libraries emerge.

I aim for coexistence with other programs and free competition.

**Comment by akondas** [comment]
@yuichiis @KEINOS thanks for the words of support. Don't get me wrong, I'm not a quit open source. I just have to complain out loud sometimes.

I admit that https://rindow.github.io/mathematics/index.html encourages me to go back to php-mlx.
Plus as a new project, we always have the advantage that we can design everything from scratch.

**Comment by torian257x** [comment]
you did advertise rubix ml by doing so. 

Now I looked at them, but am not sure if they are really better, or faster etc it isn't clear

there is a guy training something and the training took him 6 hours for something... and they say they have no GPU support.

Imho there really little reason to implement something NEW without GPU support. Or is there?

**Comment by ConnectGrid** [comment]
@akondas 

I understand how life and priorities take precedence over building and maintaining open-source projects. We do thank you and appreciate your contributions tremendously.

I'm just curious why the `php-ml` project seems to have disappeared from GitHub entirely? Am I misunderstanding something? Since this project has zero code committed, would it make sense to leave the other project online, in archived status? 

Hoping all is well.

**Comment by KEINOS** [comment]
> I'm just curious why the php-ml project seems to have disappeared from GitHub entirely? Am I misunderstanding something? Since this project has zero code committed, would it make sense to leave the other project online, in archived status?

https://arkadiuszkondas.com/dmca-php-ml-and-copyright-boundaries/

Every time I read this blog post, my heart aches and I get a glimpse into why PHP can't take the next step in the ML world and why it can't overtake Python.

**Comment by MustafaKarabulut** [comment]
It's been a while. 

I know this repo is no longer active and all people moved on to other ventures.

However, today while browsing Github, I don't know why but I ended up here. Glad I did it because after a quick glance at what have been posted, I instantly realized that @yuichiis's Rindow along with its side projects have been evolved into something amazing. It is too bad @akondas has quit this project but I am happy to see that Rindows looks like an easy and cool replacement to PHP-ML and I think it by far surpasses it in some aspects. Kudos to you mate @yuichiis!

I knew that @yuichiis was an amazing programmer when he fixed my stupid errors in a complex code I committed to PHP-ML. This praise for his was overdue and I am happy I can repay now!

Nice to know all of you guys. Hopefully see you again at somewhere.
